pkg/kubectl/rollback.go

```diff
@@ -119,113 +115,132 @@ func (r *DeploymentRollbacker) Rollback(obj runtime.Object, updatedAnnotations m
 		return "", fmt.Errorf("failed to retrieve Deployment %s: %v", name, err)
 	}
 
+	templateForRevision, latestTemplate, err := deploymentRevision(deployment, r.c, toRevision)
+	if err != nil {
+		return "", err
+	}
 	if dryRun {
-		return simpleDryRun(deployment, r.c, toRevision)
+		return printTemplate(templateForRevision)
 	}
 	if deployment.Spec.Paused {
 		return "", fmt.Errorf("you cannot rollback a paused deployment; resume it first with 'kubectl rollout resume deployment/%s' and try again", name)
 	}
-	deploymentRollback := &extensionsv1beta1.DeploymentRollback{
-		Name:               name,
-		UpdatedAnnotations: updatedAnnotations,
-		RollbackTo: extensionsv1beta1.RollbackConfig{
-			Revision: toRevision,
-		},
+
+	// Skip if the revision already matches current Deployment
+	if equalIgnoreHash(templateForRevision, latestTemplate) {
```

Comment:
> compare it against deployment's template instead of the latest template, because Deployment controller knows which RS is the latest by comparing templates

doesn't that make us vulnerable to admission-changed fields?

Reply:
We're copying RS template back to Deployment template when rolling back anyways.

Reply:
> We're copying RS template back to Deployment template when rolling back anyways.

So after the first rollback the thing in the deployment template will be something that went through admission. Seems like we'd want the first rollback to be consistent with that behavior as well, right?

Reply:
Because RS revision annotation can be changed by the users, assuming that RS with the largest revision will be the exact revision the Deployment is running is not as reliable as comparing templates directly. This behavior is also consistent with Deployment controller. Deployment controller never relies on revision annotation but only uses it as a reference and as a way to distinguish different revisions of replicasets in `kubectl rollout history`.